mlir/lib/Transforms/SROA.cpp

```diff
@@ -0,0 +1,201 @@
+    for (OpOperand *blockingUse : newBlockingUses) {
+      assert(llvm::find(user->getResults(), blockingUse->get()) !=
+             user->result_end());
+
+      SmallPtrSetImpl<OpOperand *> &newUserBlockingUseSet =
+          getOrInsertDefault(info.userToBlockingUses, blockingUse->getOwner());
+      newUserBlockingUseSet.insert(blockingUse);
+    }
+  }
+
+  return info;
+}
+
+void mlir::destructSlot(DestructibleMemorySlot &slot,
+                        DestructibleAllocationOpInterface allocator,
+                        OpBuilder &builder, MemorySlotDestructionInfo &info) {
+  OpBuilder::InsertionGuard guard(builder);
+
+  builder.setInsertionPointToStart(slot.slot.ptr.getParentBlock());
+  DenseMap<Attribute, MemorySlot> subslots =
+      allocator.destruct(slot, info.usedIndices, builder);
+
+  llvm::SetVector<Operation *> usersToRewire;
+  for (auto &[user, _] : info.userToBlockingUses)
+    usersToRewire.insert(user);
+  for (DestructibleAccessorOpInterface accessor : info.accessors)
+    usersToRewire.insert(accessor);
+  SetVector<Operation *> sortedUsersToRewire =
+      mlir::topologicalSort(usersToRewire);
+
+  llvm::SmallVector<Operation *> toErase;
+  for (Operation *toRewire : llvm::reverse(sortedUsersToRewire)) {
+    builder.setInsertionPointAfter(toRewire);
+    if (auto accessor = dyn_cast<DestructibleAccessorOpInterface>(toRewire)) {
+      if (accessor.rewire(slot, subslots) == DeletionKind::Delete)
+        toErase.push_back(accessor);
+      continue;
+    }
+
+    auto promotable = cast<PromotableOpInterface>(toRewire);
+    if (promotable.removeBlockingUses(info.userToBlockingUses[promotable],
+                                      builder) == DeletionKind::Delete)
+      toErase.push_back(promotable);
+  }
+
+  for (Operation *toEraseOp : toErase)
+    toEraseOp->erase();
+
+  assert(slot.slot.ptr.use_empty() &&
+         "at the end of destruction, the original slot "
+         "pointer should no longer be used");
+
+  allocator.handleDestructionComplete(slot);
+}
+
+namespace {
+
+struct SROA : public impl::SROABase<SROA> {
+  void runOnOperation() override {
+    Operation *scopeOp = getOperation();
+    bool changed = false;
+
+    for (Region &region : scopeOp->getRegions()) {
+      if (region.getBlocks().empty())
+        continue;
+
+      OpBuilder builder(&region.front(), region.front().begin());
+
+      // Destructing a slot can allow for further destruction of other slots,
+      // destruction is tried until no destruction succeeds.
+      while (true) {
+        struct DestructionJob {
+          DestructibleAllocationOpInterface allocator;
+          DestructibleMemorySlot slot;
+          MemorySlotDestructionInfo info;
+        };
+
+        std::vector<DestructionJob> toDestruct;
+
+        for (Block &block : region)
+          for (Operation &op : block.getOperations())
+            if (auto allocator =
+                    llvm::dyn_cast<DestructibleAllocationOpInterface>(op))
```

Comment:
usually we try to avoid too much nesting depth. Can we use `walk([&](DestructibleAllocationOpInterface allocator)`?